tests/unit/test_compress.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import gzip
+import io
+from copy import deepcopy
+
+import pytest
+
+from botocore.compat import urlencode
+from botocore.compress import COMPRESSION_MAPPING, maybe_compress_request
+from botocore.config import Config
+from tests import mock
+
+
+def _make_op(
+    request_compression=None,
+    has_streaming_input=False,
+    streaming_metadata=None,
+):
+    op = mock.Mock()
+    op.request_compression = request_compression
+    op.has_streaming_input = has_streaming_input
+    if streaming_metadata is not None:
+        streaming_shape = mock.Mock()
+        streaming_shape.metadata = streaming_metadata
+        op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+OP_NO_COMPRESSION = _make_op()
+OP_WITH_COMPRESSION = _make_op({'encodings': ['gzip']})
+OP_UNKNOWN_COMPRESSION = _make_op({'encodings': ['foo']})
+OP_MULTIPLE_COMPRESSIONS = _make_op({'encodings': ['gzip', 'foo']})
+STREAMING_OP_WITH_COMPRESSION = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {},
+)
+STREAMING_OP_WITH_COMPRESSION_REQUIRES_LENGTH = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {'requiresLength': True},
+)
+REQUEST_BODY = (
+    b'Action=PutMetricData&Version=2010-08-01&Namespace=Namespace'
+    b'&MetricData.member.1.MetricName=metric&MetricData.member.1.Unit=Bytes'
+    b'&MetricData.member.1.Value=128'
+)
+
+REQUEST_BODY_STRING = REQUEST_BODY.decode('utf-8')
+
+DEFAULT_COMPRESSION_CONFIG = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=10420,
+)
+COMPRESSION_CONFIG_128_BYTES = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=128,
+)
+COMPRESSION_CONFIG_1_BYTE = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=1,
+)
```

Comment:
128 comes from test cases outlined in the spec. I believe that number was chosen because its a sensible value that is less than the length of `REQUEST_BODY`, which is also used in the test cases in the spec. So its a useful way to outline test cases using this body derived from the `PutMetricData` operation where we're validating whether or not compression is enabled since we know its of the requisite length.

1 could be removed, but I chose it for a few reasons. First I've defined a request dict using a dict body that is rather short `{'foo': 'bar'}`. For this case specifically, I wanted to ensure that compression was enabled and able to parse/compress a dictionary body successfully. However, I could parse `REQUEST_BODY` into a dictionary instead. I also used it for test cases where we expect that the length of the body couldn't be obtained (incompatible types and nonseekable streams whose operations don't have streaming input. These are essentially testing `_get_body_size` is allowing us to disable compression for certain body types even when it is at its absolute lowest.

10240 is the default value. When using it with `REQUEST_BODY`, it is a convenient way to test inverse cases that we 're doing for 128. We know that this number is quite a lot bigger than the length of the body, so we can use it to check when we expect compression to be enabled/disabled. For example, if an operation has the compression trait and streaming input, we know that compression should occur regardless of the length of the body.

Reply:
I believe we can safely migrate away from 1 without changing any of the test cases, except for the dictionary. Let me know if you think I should go ahead and do that.